### PR TITLE
Fix OO token teleporting issue when laying green OO tile

### PR DESCRIPTION
**Pull request by michaeljb**
This breaks a couple connection tests, but I don't entirely understand what's happening in those tests; added some questions in comments in aa40d57.

Still need to make green OO tiles upgradable to brown, so I'm marking this as a Draft PR. Additionally I now seem to be unable to run a route through both cities on a OO tile.

In this flie, it is C&A's turn, they cannot upgrade J4 or run through it twice, though both should be possible.
[cannot_run_or_upgrade_oo.json.txt](https://github.com/tobymao/18xx/files/4654826/cannot_run_or_upgrade_oo.json.txt)

### CODE REVIEW

File: lib/engine/part/city.rb [hunk 1]
```diff
@@ -73,6 +74,11 @@ def place_token(corporation)
       def exchange_token(token)
         @tokens[get_slot(token.corporation)] = token
       end
+
+      # returns Array[Integer] of all edges that this city are connected to
+      def connected_edges
+        @connected_edges[tile.rotation] ||= tile.paths.select { |p| p.city == self }.flat_map(&:edges).map(&:num).sort
```
Comment (tobymao) [inline lib/engine/part/city.rb:80]: you should never call edges, just call exits

File: lib/engine/part/city.rb [hunk 2]
```diff
@@ -17,6 +17,7 @@ def initialize(revenue, slots = 1)
         @slots = slots.to_i
         @tokens = Array.new(@slots)
         @reservations = []
+        @connected_edges = {}
```
Comment (tobymao) [inline lib/engine/part/city.rb:20]: intsead of doing this, let's add to part base an attr called paths

this way every part has a back reference to it's paths

you may need to grep the code a bit to refactor this

File: lib/engine/hex.rb [hunk 3]
```diff
@@ -60,16 +60,46 @@ def name
...
+        else
+          @tile.cities.map do |old_city|
+            cities = tile.cities.select do |new_city|
+              # we want old_edges to be subset of new_edges
+              old_edges = old_city.connected_edges
+              new_edges = new_city.connected_edges
+              old_edges & new_edges == old_edges
+            end.compact
+
+            unless cities.one?
```
Comment (tobymao) [inline lib/engine/hex.rb:82]: it it possible that it maps to another city?

Comment (michaeljb) [reply lib/engine/hex.rb:82]: Probably not...got rid of this check in cc2a5d3

File: lib/engine/part/city.rb (hunk 1 above)
Comment (michaeljb) [reply lib/engine/part/city.rb:80]: d9d46a6

**Comment by michaeljb** [comment]
@tobymao I fixed the tests and removed my comments, but I haven't gone through the connection code very much yet, so I'm not fully confident my "fix" makes sense - 8a0ff24

File: lib/engine/part/city.rb (hunk 2 above)
Comment (michaeljb) [reply lib/engine/part/city.rb:20]: Given that when the parts are created, all the paths are not necessarily created (https://github.com/tobymao/18xx/blob/3d43922ace1febb16591b609272a3cd2ea36d036/lib/engine/tile.rb#L68-L104), would you be fine with the different part classes having a `paths` method that calculates `@paths` based on `@tile`?

**Comment by tobymao** [comment]
hmmm let me think about it, let's chat on slack m aybe

File: lib/engine/part/city.rb (hunk 2 above)
Comment (michaeljb) [reply lib/engine/part/city.rb:20]: For example, here's how `Engine::Part::City` could look:

```
def initialize
  # ...
  @paths = {}
end

# ...

def paths
  @paths[tile.rotation] ||= tile.paths.select { |p| p.city == self }
end

def connected_edges
  paths.flat_map(&:exits).sort
end

**Comment by tobymao** [comment]
@michaeljb it makes sense, this error was due to kelsin migrating rotations and i didn't fix the test!

**Comment by michaeljb** [comment]
> @michaeljb it makes sense, this error was due to kelsin migrating rotations and i didn't fix the test!

Perfect, that makes way more sense!

**Pull request merged**